File: 2020/11/week1/minhyungPark/algorithm/Leetcode1629.kt
/**
 * Leetcode 1629. Slowest Key
 * https://leetcode.com/contest/weekly-contest-212/problems/slowest-key/
 */
import kotlin.math.max

class Leetcode1629 {

    fun slowestKey(releaseTimes: IntArray, keysPressed: String): Char {
        val durations = mutableMapOf<Char, Int>()
        for (i in releaseTimes.indices) {
            val time = if (i == 0) releaseTimes[i] else releaseTimes[i] - releaseTimes[i-1]
            durations[keysPressed[i]] = max(durations.getOrDefault(keysPressed[i], 0), time)
        }
        return durations.map { it.key to it.value }.sortedWith(compareBy({it.second}, {it.first})).last().first
    }
}

File: 2020/11/week2/1ilsang/algorithm/leetcode1629.js
/**
 * 1629. Slowest Key
 * https://leetcode.com/contest/weekly-contest-212/problems/slowest-key/
 * @param {number[]} releaseTimes
 * @param {string} keysPressed
 * @return {character}
 */
const slowestKey = (releaseTimes, keysPressed) => {
  const initKey = keysPressed[0].charCodeAt() - 97;
  const answer = {
    key: initKey,
    value: releaseTimes[0],
  };

  for (let i = 1; i < releaseTimes.length; i++) {
    const gap = releaseTimes[i] - releaseTimes[i - 1];
    const curKey = keysPressed[i].charCodeAt() - 97;

    if (answer.value < gap) {
      answer.key = curKey;
      answer.value = gap;
    } else if (answer.value === gap && answer.key < curKey) {
      answer.key = curKey;
    }
  }

  return String.fromCharCode(answer.key + 97);
};
